src/core/meshcop/border_agent.cpp

```diff
@@ -0,0 +1,596 @@
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a Border Agent role.
+ */
+
+#define WPP_NAME "border_agent.tmh"
+
+#ifdef OPENTHREAD_CONFIG_FILE
+#include OPENTHREAD_CONFIG_FILE
+#else
+#include <openthread-config.h>
+#endif
+
+#include <stdio.h>
+#include <string.h>
+
+#include "openthread/platform/random.h"
+
+#include <coap/coap_header.hpp>
+#include <common/crc16.hpp>
+#include <common/encoding.hpp>
+#include <common/logging.hpp>
+#include <meshcop/border_agent.hpp>
+#include <meshcop/tlvs.hpp>
+#include <thread/thread_netif.hpp>
+#include <thread/thread_tlvs.hpp>
+#include <thread/thread_uris.hpp>
+
+using Thread::Encoding::BigEndian::HostSwap64;
+
+namespace Thread {
+namespace MeshCoP {
+
+BorderAgent::BorderAgent(ThreadNetif &aThreadNetif):
+    mRelayReceive(OPENTHREAD_URI_RELAY_RX, &BorderAgent::HandleRelayReceive, this),
+    mRelayTransmit(OPENTHREAD_URI_RELAY_TX, &BorderAgent::HandleRelayTransmit, this),
+    mCommissionerPetition(OPENTHREAD_URI_COMMISSIONER_PETITION, &BorderAgent::HandleCommissionerPetition, this),
+    mCommissionerKeepAlive(OPENTHREAD_URI_COMMISSIONER_KEEP_ALIVE, &BorderAgent::HandleCommisionerKeepAlive, this),
+    mMgmtCommissionerSet(OPENTHREAD_URI_COMMISSIONER_SET, &BorderAgent::HandleMgmtCommissionerSet, this),
+    mTimer(aThreadNetif.GetIp6().mTimerScheduler, &BorderAgent::HandleTimer, this),
+    mNetif(aThreadNetif)
+{
+    mNetif.GetCoapServer().AddResource(mRelayReceive);
+    mNetif.GetBASecureCoapServer().AddResource(mRelayTransmit);
+    mNetif.GetBASecureCoapServer().AddResource(mCommissionerPetition);
+    mNetif.GetBASecureCoapServer().AddResource(mCommissionerKeepAlive);
+    mNetif.GetBASecureCoapServer().AddResource(mMgmtCommissionerSet);
+}
+
+otInstance *BorderAgent::GetInstance()
+{
+    return mNetif.GetInstance();
+}
+
+ThreadError BorderAgent::Start(void)
+{
+    ThreadError error = kThreadError_None;
+    if (mNetif.GetDtls().IsStarted())
+    {
+        SuccessOrExit(error = mNetif.GetBASecureCoapServer().Stop());
+        mNetif.GetDtls().Stop();
+    }
+    mTimer.Start(kWaitChildIDResponseTime);
+exit:
+    return error;
+}
+
+ThreadError BorderAgent::Stop(void)
+{
+    ThreadError error = kThreadError_None;
+
+    otLogFuncEntry();
+
+    SuccessOrExit(error = mNetif.GetBASecureCoapServer().Stop());
+
+    mNetif.GetDtls().Stop();
+
+exit:
+    otLogFuncExitErr(error);
+    return error;
+}
+
+ThreadError CopyPayload(Message &srcMessage, Message *destMessage)
+{
+    ThreadError error = kThreadError_None;
+
+    while (srcMessage.GetOffset() < srcMessage.GetLength())
+    {
+        uint16_t length = srcMessage.GetLength() - srcMessage.GetOffset();
+        uint8_t tmp[16];
+
+        if (length >= sizeof(tmp))
```

Comment:
Should this simply be a `>` instead of a `>=`? If `length` is equal to `sizeof(tmp)` then we don't really need to make that assignment.